File: iphone-code/code/Stasher/STRegisterStepThreeViewController.m
//
//  STRegisterStepThreeViewController.m
//  Stasher
//
//  Created by bhushan on 10/11/14.
//  Copyright (c) 2014 OAB. All rights reserved.
//

#import "STRegisterStepThreeViewController.h"
#import "STBaseTabBarController.h"

@interface STRegisterStepThreeViewController ()

@end

@implementation STRegisterStepThreeViewController

- (void)viewWillAppear:(BOOL)animated{
    [super viewWillAppear:YES];
    [self.tabBarController.tabBar setHidden:YES];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self.buttonSaveLater.titleLabel setFont:[UIFont FontForButtonsWithSize:11.0f]];
    [self.buttonSaveLater setTitleColor:[UIColor stasherTextColor] forState:UIControlStateNormal];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(bankTransactionFinished:) name:kNotificationIdentifier_BankTransactionFinished object:nil];
    // Do any additional setup after loading the view.
    [self.headerLabelRegistration setText:@"Registration"];
    [self.headerLabelRegistration setFont:[UIFont FontForHeader]];
    [self.headerLabelStep setFont:[UIFont FontForHeader]];
    [self.headerLabelStep sizeToFit];
    [self.headerLabelRegistration sizeToFit];
    [self.selectBankDetailsLabel setFont:[UIFont FontForHeaderNoBold]];
    [self.selectBankDetailsLabel setTextColor:[UIColor stasherTextColor]];

    //create dictionary with user information
    NSMutableDictionary *userInfo = [[NSMutableDictionary alloc] init];
    
    //add api key
    [userInfo setObject:@"4aa327f5ac661878d6178e2c2b44aa5a9a0708d2" forKey:@"api_key"];
    
    //add api password
    [userInfo setObject:@"7a383e494ac323569402b9f1fe591c1735906202" forKey:@"api_password"];
    
    //add amount as string
    [userInfo setObject:@"0.01" forKey:@"amount"];
    //[userInfo setObject:@"0.01" forKey:@"amount"];
    
    //add invoice detail];
    [userInfo setObject:@"Registration Token Transaction" forKey:@"invoice_detail"];
    //[userInfo setObject:@"invoice detail" forKey:@"invoice_detail"];
    
    //add information request
    [userInfo setObject:@"show_all" forKey:@"information_request"];
    
    //add recurring detail
    [userInfo setObject:@"ot" forKey:@"reccuring"];
    
    //add redirect URL
    [userInfo setObject:@"fb743090835784071://" forKey:@"redirect_url"];
    
    //make url string with user info on the end
    NSString *url = [NSString stringWithFormat:@"https://knoxpayments.com/pay?api_key=%@&api_password=%@&amount=%@&invoice_detail=%@&information_request=%@&recurring=%@&redirect_url=%@",[userInfo objectForKey:@"api_key"],[userInfo objectForKey:@"api_password"],[userInfo objectForKey:@"amount"],[userInfo objectForKey:@"invoice_detail"],[userInfo objectForKey:@"information_request"],[userInfo objectForKey:@"reccuring"],[userInfo objectForKey:@"redirect_url"] ];
    url = [url stringByAddingPercentEscapesUsingEncoding:NSASCIIStringEncoding];
    NSLog(@"%@",url);
    
    //create URL and request
    NSURL *nsurl = [NSURL URLWithString:url];
    NSURLRequest *nsrequest = [NSURLRequest requestWithURL:nsurl];
    [self.paymentsWebView loadRequest:nsrequest];
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

- (IBAction)backButtonPressed:(id)sender
{
    [self.navigationController popViewControllerAnimated:YES];
}

- (IBAction)saveDetailsButtonPressed:(id)sender
{
    UIStoryboard* storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:[NSBundle mainBundle]];
    STBaseTabBarController *baseTabsController = [storyboard instantiateViewControllerWithIdentifier:@"STBaseTabBarController"];
    [self.navigationController pushViewController:baseTabsController animated:YES];
}

- (IBAction)saveLaterButtonPressed:(id)sender
{
    UIStoryboard* storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:[NSBundle mainBundle]];
    STBaseTabBarController *baseTabsController = [storyboard instantiateViewControllerWithIdentifier:@"STBaseTabBarController"];
    [self.navigationController pushViewController:baseTabsController animated:YES];
}

- (void)bankTransactionFinished:(NSNotification*)notification
{
    NSDictionary *dict = [notification userInfo];
    if (dict) {
        if ([dict objectForKey:@"pay_id"]) {
            [[STUserIdentity sharedInstance] setUserKnoxFirstTransactionID:[dict objectForKey:@"pay_id"]];
            if (httpReq) {
                httpReq.delegate = nil;
                httpReq = nil;
            }
            httpReq = [[BOABHttpReq alloc] initBoabReqWithDelegate:self shouldShowActivityIndicatorView:YES];
            NSString *userIdString, *apiActionString;
            userIdString = kParamKeyParentID;
            apiActionString = kAPIActionAddBankAccount;
            [httpReq sendAsyncPostRequest:STASHER_BASE_URL params:[NSDictionary dictionaryWithObjectsAndKeys:[NSNumber numberWithInt:[[[STUserIdentity sharedInstance] userId] intValue]], userIdString,apiActionString,kParamKeyAction,[dict objectForKey:@"pay_id"],kParamKeyTransactionID,nil] json:YES];
        }
        if ([dict objectForKey:@"completed"]) {
            
        }
    }
    [self saveLaterButtonPressed:nil];
    [[NSNotificationCenter defaultCenter] removeObserver:self name:kNotificationIdentifier_BankTransactionFinished object:nil];
}


# pragma mark ----- Textfield Methods
- (BOOL)textFieldShouldReturn:(UITextField *)textField
{
    [textField resignFirstResponder];
    return YES;
}

#pragma mark ----- WebView Delegate

- (void)webViewDidStartLoad:(UIWebView *)webView
{
    if (self.activityIndicatorView!=nil) {
        [self.activityIndicatorView removeFromSuperview];
        self.activityIndicatorView = nil;
    }
    self.activityIndicatorView = [[STActivityIndicatorView alloc] initWithDefaultSizeandSuperView:self.view];
    [self.activityIndicatorView startAnimation];
}

- (void)webViewDidFinishLoad:(UIWebView *)webView
{
    [self.activityIndicatorView stopAnimation];
}

- (void)webView:(UIWebView *)webView didFailLoadWithError:(NSError *)error
{
    [self.activityIndicatorView stopAnimation];
    if (self.activityIndicatorView!=nil) {
        [self.activityIndicatorView removeFromSuperview];
        self.activityIndicatorView = nil;
    }
}

#pragma mark ----- BOABHttpReqDelegates

- (void)BOABHttpReqFinishedWithResponse:(NSData*)resonseData andConnection:(BOABURLConnection*)conn
{
    if (resonseData) {
        NSDictionary *responseDictionary = [NSJSONSerialization JSONObjectWithData:resonseData options:kNilOptions error:nil];
        if (responseDictionary) {
            NSLog(@"transaction_id sent to server...");
        }
    }
}


@end

File: swift-code/Stasher/STRegisterStepThreeViewController.swift
//  Converted with Swiftify v1.0.6292 - https://objectivec2swift.com/
//
//  STRegisterStepThreeViewController.swift
//  Stasher
//
//  Created by bhushan on 10/11/14.
//  Copyright (c) 2014 OAB. All rights reserved.
//
import UIKit
class STRegisterStepThreeViewController: UIViewController, BOABHttpReqDelegate {
    var httpReq: BOABHttpReq?

    @IBOutlet var textFieldCardNumber: UITextField!
    @IBOutlet var textFieldExpiryDate: UITextField!
    @IBOutlet var textFieldCVVNumber: UITextField!
    @IBOutlet var buttonSaveDetails: UIButton!
    @IBOutlet var buttonSaveLater: UIButton!
    @IBOutlet var paymentsWebView: UIWebView!
    @IBOutlet var selectBankDetailsLabel: UILabel!
    var activityIndicatorView: STActivityIndicatorView?
    @IBOutlet var headerLabelRegistration: UILabel!
    @IBOutlet var headerLabelStep: UILabel!


    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(true)
        tabBarController?.tabBar?.isHidden = true
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        buttonSaveLater.titleLabel?.font = UIFont(size: 11.0)
        buttonSaveLater.setTitleColor(UIColor.stasherText(), for: .normal)
        NotificationCenter.default.addObserver(self, selector: #selector(self.bankTransactionFinished), name: kNotificationIdentifier_BankTransactionFinished, object: nil)
        // Do any additional setup after loading the view.
        headerLabelRegistration.text = "Registration"
        headerLabelRegistration.font = UIFont()
        headerLabelStep.font = UIFont()
        headerLabelStep.sizeToFit()
        headerLabelRegistration.sizeToFit()
        selectBankDetailsLabel.font = UIFont()
        selectBankDetailsLabel.textColor = UIColor.stasherText()
            //create dictionary with user information
        var userInfo = [AnyHashable: Any]()
        //add api key
        userInfo["api_key"] = "4aa327f5ac661878d6178e2c2b44aa5a9a0708d2"
        //add api password
        userInfo["api_password"] = "7a383e494ac323569402b9f1fe591c1735906202"
        //add amount as string
        userInfo["amount"] = "0.01"
        //[userInfo setObject:@"0.01" forKey:@"amount"];
        //add invoice detail];
        userInfo["invoice_detail"] = "Registration Token Transaction"
        //[userInfo setObject:@"invoice detail" forKey:@"invoice_detail"];
        //add information request
        userInfo["information_request"] = "show_all"
        //add recurring detail
        userInfo["reccuring"] = "ot"
        //add redirect URL
        userInfo["redirect_url"] = "fb743090835784071://"
            //make url string with user info on the end
        var url: String? = "https://knoxpayments.com/pay?api_key=\(userInfo["api_key"] as? String)&api_password=\(userInfo["api_password"] as? String)&amount=\(userInfo["amount"] as? String)&invoice_detail=\(userInfo["invoice_detail"] as? String)&information_request=\(userInfo["information_request"] as? String)&recurring=\(userInfo["reccuring"] as? String)&redirect_url=\(userInfo["redirect_url"] as? String)"
        url = url?.addingPercentEscapes(using: String.Encoding.ascii)
        print("\(url)")
            //create URL and request
        var nsurl = URL(string: url)
        var nsrequest = URLRequest(url: nsurl!)
        paymentsWebView.loadRequest(nsrequest)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    /*
    #pragma mark - Navigation
    
    // In a storyboard-based application, you will often want to do a little preparation before navigation
    - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
        // Get the new view controller using [segue destinationViewController].
        // Pass the selected object to the new view controller.
    }
    */

    @IBAction func backButtonPressed(_ sender: Any) {
        navigationController?.popViewController(animated: true)
    }

    @IBAction func saveDetailsButtonPressed(_ sender: Any) {
        var storyboard = UIStoryboard(name: "Main", bundle: Bundle.main)
        var baseTabsController: STBaseTabBarController? = storyboard.instantiateViewController(withIdentifier: "STBaseTabBarController")
        navigationController?.pushViewController(baseTabsController, animated: true)
    }

    @IBAction func saveLaterButtonPressed(_ sender: Any) {
        var storyboard = UIStoryboard(name: "Main", bundle: Bundle.main)
        var baseTabsController: STBaseTabBarController? = storyboard.instantiateViewController(withIdentifier: "STBaseTabBarController")
        navigationController?.pushViewController(baseTabsController, animated: true)
    }

    func bankTransactionFinished(_ notification: Notification) {
        var dict: [AnyHashable: Any]? = notification.userInfo
        if dict != nil {
            if dict?["pay_id"] {
                STUserIdentity.sharedInstance().userKnoxFirstTransactionID = dict?["pay_id"]
                if httpReq != nil {
                    httpReq?.delegate = nil
                    httpReq = nil
                }
                httpReq = BOABHttpReq.initBoabReq(withDelegate: self, shouldShowActivityIndicatorView: true)
                var userIdString: String
                var apiActionString: String
                userIdString = kParamKeyParentID
                apiActionString = kAPIActionAddBankAccount
                httpReq?.sendAsyncPostRequest(STASHER_BASE_URL, params: [
                    userIdString : Int(CInt(STUserIdentity.sharedInstance().userId())),
                    kParamKeyAction : apiActionString,
                    kParamKeyTransactionID : dict?["pay_id"]
                ]
, json: true)
            }
            if dict?["completed"] {

            }
        }
        saveLaterButtonPressed(nil)
        NotificationCenter.default.removeObserver(self, name: kNotificationIdentifier_BankTransactionFinished, object: nil)
    }
// MARK: ----- Textfield Methods

    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        textField.resignFirstResponder()
        return true
    }
// MARK: ----- WebView Delegate

    func webViewDidStartLoad(_ webView: UIWebView) {
        if activityIndicatorView != nil {
            activityIndicatorView?.removeFromSuperview()
            activityIndicatorView = nil
        }
        activityIndicatorView = STActivityIndicatorView(defaultSizeandSuperView: view)
        activityIndicatorView?.startAnimation()
    }

    func webViewDidFinishLoad(_ webView: UIWebView) {
        activityIndicatorView?.stopAnimation()
    }

    func webView(_ webView: UIWebView, didFailLoadWithError error: Error) {
        activityIndicatorView?.stopAnimation()
        if activityIndicatorView != nil {
            activityIndicatorView?.removeFromSuperview()
            activityIndicatorView = nil
        }
    }
// MARK: ----- BOABHttpReqDelegates

    func boabHttpReqFinished(withResponse resonseData: Data, andConnection conn: BOABURLConnection) {
        if resonseData {
            var responseDictionary: [AnyHashable: Any]? = try? JSONSerialization.jsonObject(withData: resonseData, options: kNilOptions)
            if responseDictionary != nil {
                print("transaction_id sent to server...")
            }
        }
    }
}